Speculative improvements for rdar://125652518

beidson opened the pull request:
#### ca760255d2d5e4dffe3e11b5b342e58dc28e1d4b
<pre>
Speculative improvements for <a href="https://rdar.apple.com/125652518">rdar://125652518</a>
<a href="https://rdar.apple.com/125652518">rdar://125652518</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=272201">https://bugs.webkit.org/show_bug.cgi?id=272201</a>

Reviewed by Per Arne Vollan and Ryosuke Niwa.

While we continue to explore this crash tracer (while deserializing a postMessage to a script message handler),
I spotted some room for improvements.

* Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
(API::SharedJSContext::ensureContext):
(API::sharedContext):
(API::SerializedScriptValue::deserialize):
(API::coreValueFromNSObject):
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:

Canonical link: <a href="https://commits.webkit.org/277131@main">https://commits.webkit.org/277131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63a53f7d7363779acbfc78ebbc0da90731c3e7cd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46765 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25926 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49388 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49442 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42812 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30316 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23390 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47345 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22924 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19388 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4811 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43013 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41779 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51314 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21774 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45382 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23062 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10337 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/22770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pvollan reviewed (approved):
R=me.

pvollan commented on Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
```diff
@@ -88,20 +88,21 @@ void releaseContextIfNecessary()
         return nil;
 
-    JSValue *value = [JSValue valueWithJSValueRef:valueRef inContext:context];
+    JSValue *value = [JSValue valueWithJSValueRef:valueRef inContext:context.get()];
```
Calling 'get()' here is perhaps not strictly needed?

cdumez replied:
As far as I know, it is needed

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
```diff
@@ -88,20 +88,21 @@ void releaseContextIfNecessary()
 
 static SharedJSContext& sharedContext()
 {
+    ASSERT(RunLoop::isMain());
```
We should just use `MainThreadNeverDestroyed` below then.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
```diff
@@ -88,20 +88,21 @@ void releaseContextIfNecessary()
-    JSContext* context = sharedContext().ensureContext();
+    RetainPtr context = sharedContext().ensureContext();
 
     JSValueRef valueRef = serializedScriptValue.deserialize([context JSGlobalContextRef], exception);
```
Shouldn't we have a ref of the stack (`JSRetainPtr`) to the value returned by `[context JSGlobalContextRef]`?

beidson replied:
As long as we're hardening with smart pointers, smart point all the things! (added)

beidson committed:
Speculative improvements for rdar://125652518
rdar://125652518
https://bugs.webkit.org/show_bug.cgi?id=272201

Reviewed by Per Arne Vollan and Ryosuke Niwa.

While we continue to explore this crash tracer (while deserializing a postMessage to a script message handler),
I spotted some room for improvements.

* Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
(API::SharedJSContext::ensureContext):
(API::sharedContext):
(API::SerializedScriptValue::deserialize):
(API::coreValueFromNSObject):
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:

Canonical link: https://commits.webkit.org/277131@main
